blog/2020-03-01-CombiningBPMNWithOptaplanner.adoc

```diff
@@ -0,0 +1,611 @@
+                seat.type = (column == 0 || column == columnCount - 1)? Seat.Type.WINDOW :
+                    (column == MIDDLE_OF_ROW || column == MIDDLE_OF_ROW + 1)? Seat.Type.AISLE :
+                        Seat.Type.OTHER;
+                out.add(seat);
+            }
+        }
+        return out;
+    }
+}
+----
+
+Now that we have our domain model, we can create our services. For the
+`addPassengerToFlight` service, we can simply add the passenger to the
+flight's passenger list. For the `createSeatAssignments` service,
+we can use `SolverManager` to create a `Solver` for us and wait for the
+final best solution.
+
+[source,java]
+----
+@ApplicationScoped
+@Component
+public class FlightService {
+    @Autowired
+    private SolverManager<Flight, String> solverManager;
+
+    public void addPassengerToFlight(Flight flight, Passenger passenger) {
+        flight.passengerList.add(passenger);
+    }
+
+    public Flight createSeatAssignments(Flight flight) {
+        try {
+            return solverManager.solve(flight.id, flight).getFinalBestSolution();
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+}
+----
+
+Finally, we need to modify `application.properties` so the `Solver` know
+when to terminate. Modify your `application.properties` to look like the following:
+
+[source,sh]
+----
+server.address=0.0.0.0
+
+spring.mvc.servlet.path=/docs
+
+resteasy.jaxrs.scan-packages=org.kie.**,com.company,http*
+
+# Solve for 5 seconds
+optaplanner.solver.termination.spent-limit=5s
+----
+
+== Creating the Constraints
+
+We can use the Constraint Stream API to create the constraints of our
+application in Java. We'll have two constraints:
+
+* No two passengers can be in the same seat.
+* Maximize the number of passengers who get seats they prefer.
+
+To do this, we create a `ConstraintProvider` that implement these constraints:
+
+[source,java]
+----
+public class FlightSeatingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+            onlyOnePassengerPerSeat(constraintFactory),
+            seatTypePreferrence(constraintFactory)
+        };
+    }
+
+    private Constraint onlyOnePassengerPerSeat(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUniquePair(Passenger.class, Joiners.equal(p -> p.seat))
+                   .penalize("Only one passenger per seat", HardSoftScore.ONE_HARD);
+    }
+
+    private Constraint seatTypePreferrence(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Passenger.class)
+                   .join(Seat.class, Joiners.equal(p -> p.seat, s -> s))
+                   .filter((p,s) -> p.preferredSeatType != null && !p.preferredSeatType.equals(s.type))
```

Comment:
Maybe simplify to `p.preferredSeat != s.type` (enums can be compared using `==`).